### PR TITLE
use reflections to determine the request type and response type

### PR DESCRIPTION
**Pull request by truncs**
cc @willsalz @brandoniles

### CODE REVIEW

File: tchannel-core/src/main/java/com/uber/tchannel/util/ReflectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        } else if (type instanceof GenericArrayType) {
+            Type componentType = ((GenericArrayType) type).getGenericComponentType();
+            Class<?> componentClass = getClass(componentType);
+            if (componentClass != null) {
+                return Array.newInstance(componentClass, 0).getClass();
+            } else {
+                return null;
+            }
+        } else {
+            return null;
```
Comment (brandoniles) [inline tchannel-core/src/main/java/com/uber/tchannel/util/ReflectionUtils.java]: What's the effect of returning null here? Will a client have a good idea what's going on if this happens?

Comment (truncs) [inline tchannel-core/src/main/java/com/uber/tchannel/util/ReflectionUtils.java]: it should never return a null if the POJOs are specified in RequestHandler which is enforced by our class signature.

**Comment by willsalz** [comment]
LGTM. Although, There has got to a better way where we don't need to support this ourselves. I've checked out the Apache Commons - Lang package and might have what we need. I can play around ( later… ) and see if we can move this logic out of TChannel-Java

**Comment by truncs** [comment]
@willsalz Gauva's TypeToken would do the same thing I suppose, https://github.com/google/guava/wiki/ReflectionExplained#typetoken

In any case this break the API since this as RequestHandler as abstract class, so it would be nice if we can merge this and figure out the implementation details later ?

**Comment by brandoniles** [comment]
Instead of converting the interface to an abstract class, you can consider adding the abstract class below the interface as a skeletal implementation. This could allow for more flexibility in the API.

**Comment by willsalz** [comment]
+1 to a `DefaultRequestHandler` Abstract implementation of the RequestHandler.

**Comment by willsalz** [comment]
@truncs any new thoughts on @brandoniles and my comments r.e. a Default ABC + Keeping the Interface?

**Pull request merged**